src/Spago/Config.hs

```diff
@@ -120,30 +121,38 @@ parseConfig = do
 
   ConfigPath path <- view (the @ConfigPath)
   expr <- liftIO $ Dhall.inputExpr $ "./" <> path
-  case expr of
-    Dhall.RecordLit ks' -> do
-      let ks = Dhall.extractRecordValues ks'
-      let sourcesType  = Dhall.list (Dhall.auto :: Dhall.Decoder SourcePath)
-      name              <- Dhall.requireTypedKey ks "name" Dhall.strictText
-      dependencies      <- Dhall.requireTypedKey ks "dependencies" dependenciesType
-      configSourcePaths <- Dhall.requireTypedKey ks "sources" sourcesType
-      alternateBackend  <- Dhall.maybeTypedKey ks "backend" Dhall.strictText
-
-      let ensurePublishConfig = do
-            publishLicense    <- Dhall.requireTypedKey ks "license" Dhall.strictText
-            publishRepository <- Dhall.requireTypedKey ks "repository" Dhall.strictText
-            pure PublishConfig{..}
-      publishConfig <- try ensurePublishConfig
-
-      packageSet <- Dhall.requireKey ks "packages" (\case
-        Dhall.RecordLit pkgs -> parsePackageSet (Dhall.extractRecordValues pkgs)
-        something            -> throwM $ Dhall.PackagesIsNotRecord something)
-
-      pure Config{..}
-    _ -> case Dhall.TypeCheck.typeOf expr of
+  mbConfig <- parseConfig' expr
+  case mbConfig of
+    Just config -> pure config
+    Nothing -> case Dhall.TypeCheck.typeOf expr of
       Right e  -> throwM $ Dhall.ConfigIsNotRecord e
       Left err -> throwM err
 
+parseConfig'
```

Comment:
Basically, to verify that adding new packages to the raw expression worked, I parse the returned expression to see if it produced a `Config` value. To ensure I'm using the same logic as the original parser, I moved the parsing logic in `parseConfig` to `parseConfig'`. I've since renamed `parseConfig'` to `parseConfigDhallExpression`. I've also added docs to `parseConfigDhallExpression`.

So, `parseConfig` will get the normalized expression from the `./spago.dhall` file and then pass that expression into `parseConfigDhallExpression` whereas `addDependencies` will get the modified Dhall expression that includes new packages, normalizes it, and passes it into `parseConfigDhallExpression` to verify that the modification produces a valid `Config` value.